### ISSUE TITLE
Allow to scope settings by platform

### ISSUE THREAD

**Issue by jarod51** [Microsoft/vscode#5595]
Hi

I develop on 3 different platform. When synchronizing settings, snippets and so on, i often must change path, adjust font-size, etc...

So, it could be great if we had a per platform settings set (Windows, Mac, Unix)

**Comment by dbaeumer** [comment]
@bpasero, @aeschli  FYI.

**Comment by bpasero** [comment]
https://github.com/Microsoft/vscode/issues/8962

**Comment by Tyriar** [comment]
#8962 is about keybindings, not settings

**Comment by bbkane** [comment]
Did this happen? I'd like to consolidate my VSCode settings.json

**Comment by Tyriar** [comment]
@bbkane it got pushed out unfortunately, the issue will be closed with a milestone set (the version it'll land in) when it's done.

**Comment by thernstig** [comment]
I know you are working hard on all kind of nice features, still, is there any updates on when this might get a new milestone/target release?

**Comment by chparimi** [comment]
When will this feature release?

**Comment by alanlivio** [comment]
When will this feature release? [2]

**Comment by sandy081** [comment]
Its still under discussion, one approach we are thinking is to support Machine specific settings rather than Platform specific. Because most of these settings are path related and they are scoped to a machine rather to platform.

**Comment by chris400** [comment]
Plattform specific > machine specific
I want to place settings on a cloud drive and access it from various machines. In that scenario I would barely benefit from machine specific.

**Comment by sandy081** [comment]
@chris400 Can you please give me the example of those settings?

**Comment by eberkund** [comment]
@sandy081 For example, I would like to be able to use a different font on Windows vs Mac OS. Or on Linux I want to use the native title-bar.

I am using the [Setting Sync](https://marketplace.visualstudio.com/items?itemName=Shan.code-settings-sync) extension so it would be useful if I could share one configuration across all my machines.

**Comment by chris400** [comment]
@sandy081 https://github.com/Microsoft/vscode/issues/17619 describes it perfectly

**Comment by FelikZ** [comment]
There is [language specific settings](https://vscode.readthedocs.io/en/latest/getstarted/settings/#language-specific-editor-settings), which allows the following:

```json
{
  "[typescript]": {
    "editor.formatOnSave": true,
    "editor.formatOnPaste": true
  },
  "[markdown]": {
    "editor.formatOnSave": true,
    "editor.wordwrap": "on",
    "editor.renderWhitespace": "all",
    "editor.acceptSuggestionOnEnter": false
  }
}
```

I suggest to follow up this and allow something like:

```json
{
  "[[osx]]": {
    "editor.fontSize": 13,
    "terminal.integrated.fontSize": 13,
    "markdown.preview.fontSize": 13,
    "restructuredtext.preview.fontSize": 13
  },

  "{windows}": {
    "editor.fontSize": 12,
    "terminal.integrated.fontSize": 12
    "markdown.preview.fontSize": 12
    "restructuredtext.preview.fontSize": 12
  },

  "(linux)": {
    "editor.fontSize": 11,
    "terminal.integrated.fontSize": 11
    "markdown.preview.fontSize": 11
    "restructuredtext.preview.fontSize": 11
  },

  "editor.fontSize": 12,
  "terminal.integrated.fontSize": 12
  "markdown.preview.fontSize": 12
  "restructuredtext.preview.fontSize": 12
}
```

**Comment by jonnyasmar** [comment]
@FelikZ Is there a reason you formatted osx, windows, & linux keys differently? Or just illustrating possibilities? I think single brackets are most consistent: `[windows]`

@sandy081 Platform & machine specific would be great. I'm assuming some pattern like this?

```json
"[windows]: {
  "editor.fontSize": 13,
},
"[Jonny-PC]": {
  "local-history.path": "c:\dev\.vscode",
}
```

One additional consideration is that `terminal.integrated.shell` (maybe others?) follows a divergent pattern:

```json
  "terminal.integrated.shell.osx": "zsh",
  "terminal.integrated.shell.windows": "C:\\WINDOWS\\System32\\wsl.exe"
```

If platform specific settings become an option, this should probably be adjusted to follow the same pattern:

```json
"[windows]: {
  "terminal.integrated.shell": "C:\\WINDOWS\\System32\\wsl.exe"
},
"[osx]: {
  "terminal.integrated.shell": "zsh",
},
```

**Comment by bbkane** [comment]
I like the idea of using `[]` to represent conditions in JSON. Because `[language]` is already being used as a specialization option, we should consider guarding other names (I'm using sub-maps and the "platform" namespace below"). I also think it would be nice to make the override order explicit.

A theoretical example extending the namespacing with a new key `platform` and use maps inside to specialize:

```json
{
  // settings later in the list override settings earlier in the list
  "settingsOverrideOrder": [
    "default",  // stuff in settings.json not in a [condition] specialization
    "platform.os",       // settings.json in the "platform.os" map
    "platform.hostname", // settings.json in the "platform.hostnames" map
    "language", // stuff in the "[lang]" maps
    "project"   // stuff in .vscode/settings.json
  ]

  // We can't do a "language" map cause we need backwards compatibility
  "[elm]": {
    "editor.formatOnSave": true,
  },

  "[platform.os]": {
    "osx": {
      "editor.formatOnSave": true,
    }
  },

  "[platform.hostname]": {
    "mac01": {
      "editor.formatOnSave": true,
    }
    "lin01": {
      "editor.formatOnSave": true,
    }
  },

}
```

## Notes

Instead of scoping by dictionary, we could scope specializations with dots more. This would result in the "hostname" key having dynamically scoped names, which is not great... but it would reduce nesting.

```
"[platform.os.osx]": ...
"[platform.hostname.mac01]" ... // ewww...
```

---

We can't have a `"[language]": { "elm": ... }` specialization because that breaks backwards compatibility. Languages are already represented in "[lang-name]".

---

There's probably better names for a lot of this stuff... "osx" vs "darwin", "settingsOverrideOrder".

"hostname" vs "machineName" vs some other machine identifier (MAC address, IP, ...)

---

I'm namespacing new non-language specializations with dots to avoid language name conflicts (hope there's not a new language named "platform")

---

I'm introducing "settingsOverrideOrder" because I think the user might want to decide override order (particularly in regards to where language fits in).

**Comment by FelikZ** [comment]
@jonnyasmar for illustration only.

**Comment by a20012251** [comment]
I work at Oculus/Facebook and this might be the only feature missing for us to start using VSCode heavily

**Comment by chris400** [comment]
> @chris400 Can you please give me the example of those settings?

I am extending my explanation https://github.com/Microsoft/vscode/issues/5595#issuecomment-459027931: my code sits on a local server, the pathes to the code differs on every plattform.

**Comment by mazza** [comment]
+1 to desirable feature

A portable/embedded version of Python was compiled to each of those archs In a project I'm working on. The portable environment is used to both run/test the bundled portable software and provide the portable dev env as well. But this missing feature may affect any project intended to be coded in more than one OS.

```
// settings.json
{
    "terminal.integrated.env.linux": {
        "PATH": ".tools/linux/bin:${env:PATH}"
    },
    "terminal.integrated.env.osx": {
        "PATH": ".tools/macos/bin:${env:PATH}"
    },
    "terminal.integrated.env.windows": {
        "PATH": ".tools\\windows;${env:PATH}"
    },
    "python.pythonPath": ".tools/linux/bin/python3", // (?)
    "python.linting.mypyEnabled": true,
    // "python.linting.mypyPath": ".tools/linux/bin/mypy",
    "python.linting.enabled": true,
    // "python.linting.banditPath": ".tools/linux/bin/bandit",
    "python.linting.banditEnabled": true
}
```

The commented lines are OK to skip, but if "python.pythonPath" is omitted, VSCode won't find the debuggers and linters. But if it is present it may only work in one OS requiring fix at every time.

The specific way I would suggest in order to require the minimal move to get this done would be to add one more path to settings.json for each arch, so:

- settings.windows.json for Windows
- settings.linux.json for Linux
- settings.osx.json for macOS
- settings.json for all

Then, the OS-specific config file would load after the generic one completing or even overwriting it when convenient. This solution may avoid dealing with infinite possible OS-specific config issues that may arise with many new local solutions.

**Comment by rossipedia** [comment]
I don't know if this has been said, but platform specific settings are useful for more things than just paths and keybinds. I move between several boxes daily, on all 3 main platforms (Windows, Mac, Linux), and each one needs separate `window.zoomLevel` settings in order for things to look right.

**Comment by RolfNoot** [comment]
I have this for example in my settings:
"cmake.cmakePath": "/Applications/CMake.app/Contents/bin/cmake",
"cortex-debug.openocdPath": "/Applications/ModusToolbox_1.1/tools/openocd-1.0/bin/openocd",

I understand there are issues with some proposed solutions and instead of developers rejecting possible solutions, are there any developers who can propose a solution for this issue?

We're in 2019 and still have to edit settings to switch platforms?

**Comment by xahon** [comment]
suka

**Comment by RuBAN-GT** [comment]
@xahon, derzhis, bratishka, ty ne odin

**Comment by repentsinner** [comment]
Following here from my previous use case regarding python paths on different OSes (https://github.com/Microsoft/vscode/issues/17619#issuecomment-365521276), I also just discovered Shan Khan's excellent [Settings Sync](https://marketplace.visualstudio.com/items?itemName=Shan.code-settings-sync) and wish that I could keep one sync'd non-workspace `settings.json` file with separate UI configs for the different machines I work on.

An appropriate `"editor.fontSize"` is very different on a 13" MacBook vs a 32" 4K Windows machine due to the different display sizes, OS HiDPI scaling methods, font handling, etc. 

Having per-host settings would also be awesome, but per-platform settings would get us a good piece of the way there.

Given that Settings Sync has ~1M installs, it seems likely that a good number of people are syncing cross-platform, and certainly cross-host.

Thanks for considering!

**Comment by rossipedia** [comment]
@repentsinner I just looked into this again, and it seems as though Settings Sync [allows for selectively ignoring certain settings](https://github.com/shanalikhan/code-settings-sync/wiki/Sync-Pragmas#ignoring-certain-settings):

``` json
// settings.json

{
  // @sync-ignore
  "window.zoomLevel": "1", /* won't upload to gist */
}
```

I'm not sure that suits your purposes, but I think it'll get me most of the way to where I need to go

**Comment by michaeltarleton** [comment]
@rossipedia I used your link and noticed you can do per-host, os, env with Settings Sync:

```
// settings.json on linux

{
  // @sync os=linux
  "window.zoomLevel": "1"
}
```

```
// settings.json on a computer with the environmental variable "CODE_ZOOM_1" set

{
  // @sync env=CODE_ZOOM_1
  "window.zoomLevel": "1"
}
```

```
// settings.json on "home" computer

{
  // @sync host=home
  "window.zoomLevel": "1"
}
```

**Comment by jcklpe** [comment]
I'm honestly kind of surprised that this is still an issue. 

The setting sync isn't reliable enough an answer for me. I've had setting syncs revert my settings to months older versions multiple times and I don't consider it reliable enough to solve this problem. This really should be built in. 

I mean it's even possible for me to sync my zshrc settings pretty easily as both zsh and bash have really easy OS specific conditional statements possible.

**Comment by repentsinner** [comment]
Thanks @rossipedia and @michaeltarleton for the heads up on the selective sync stuff.

`@sync-ignore` does help, although I'd really like everything sync'd and then selectively _applied_. That might be closer to what Michael found, with everything landing in the gist and then selectively syncing back out to the different VS Code instances. In this case, editing the settings in the gist may effectively result in scoping settings by platform/host/... as only the appropriate settings are sync'd to an instance.

Note that there's some recent activity in https://github.com/microsoft/vscode/issues/78869 that is proposing to move sync into the VS Code project natively.

I didn't intend to derail this issue with sync vs. scoping, and believe they are two separate but substantially related topics. For example, we may also want to use scoped workspace/project settings, where in that case the sync mechanism is git or some other SCM rather than native/extension-based VS Code mechanism.

**Comment by xahon** [comment]
What the state of this issue?

**Comment by hpgmiskin** [comment]
We have started using the [platform-settings](https://marketplace.visualstudio.com/items?itemName=runarsf.platform-settings) extension for VSCode in order to get different settings for people using different computers. From first testing it seems to work well. We use it to direct to different versions of `clang-format`.

**Comment by sandy081** [comment]
This feature has to come to discussions several times and was not considered because it is mainly associated to settings sync. It means, the main use case for this feature is to enable users working on different platforms/machines to share settings per platform. I already started working on providing settings sync feature out of the box in VS Code - https://github.com/microsoft/vscode/issues/2743.  So, I also started thinking how to support this. I see two suggestions from the community here:

1. Use language specific settings style. This is thought about and after thinking further we decided not to go with this approach. Reason being language overrides has strong precedence and they can always win against user/workspace/folder settings. This is not intended for platform specific settings. See https://github.com/microsoft/vscode/issues/86737 for more information.

2. Separate platform specific files as mentioned in comment https://github.com/microsoft/vscode/issues/5595#issuecomment-485877346. This way sync will also sync settings appropriately per platform.

**Comment by CodyWalraven** [comment]
I really like idea 2, the ability to still have the main settings.json be the base with platform specific files on top that can override just the features you want seems straightforward and simple to use. Plus I think it will be less messy since you are not mixing up all of the different file paths (I'm looking at you windows with your \'s everywhere) in one file.

**Comment by sandorex** [comment]
The 2nd idea is cleaner IMO and wont add any new syntax to existing settings.json

**Comment by xahon** [comment]
2020

**Comment by yulei900609** [comment]
I need this feather very much.

**Comment by codecalvin** [comment]
Looking forward to this feature.
Like 2nd idea.

**Comment by Ovyerus** [comment]
Seeing how Insiders now has a settings/extensions sync built in without needing another extension to do it, this would be pretty handy to have now

**Comment by LeoJhonSong** [comment]
Since I am sharing workspace level settings with my members on several platforms, I really need exactly this feature but not a workaround by Settings Sync...

So do we have any news on this feature?

**Comment by crscillitoe** [comment]
I use different fonts on windows/mac/linux (their font names are slightly different, but they're all the same font).

I have no reliable way to configure the name for each OS right now, can we please get some traction on this feature?

**Comment by DeDerpster** [comment]
Just been looking through the new documentation for the settings sync in the latest update and came across [this](https://code.visualstudio.com/docs/editor/settings-sync#_configuring-synced-data), so following on @rossipedia [comment](https://github.com/microsoft/vscode/issues/5595#issuecomment-520098764), it looks like you can do something similar with the new inbuilt setting sync, by using "settingsSync.ignoredSettings" - still slightly annoying if you are constantly using new computers, though, as I think you would have to continuously update these settings every time. However I think it gets rid of a lot of issues.

**Comment by plzhans** [comment]
Looking forward to this feature. Like 2nd idea.

**Comment by juarezr** [comment]
One question:

This will allow one to have specific settings when switching between local machine and a [Remote Container](https://code.visualstudio.com/docs/remote/containers)?

What happens when the same `.vscode/settings.json` is used used in `local mode` and mapped to a remote container by a `devcontainer.json` if both run on `linux`?

**Comment by Natetronn** [comment]
Would love to a features like this! 

Setting the python interpreter on Linux and Windows over and over again isn't much fun.

**Comment by celynwalters** [comment]
I have a temporary solution using the new settings sync (which I love) until this is implemented:
1. Put the setting which is platform-dependent into `settingsSync.ignoredSettings`.
2. Comments _are_ synced, so next to the ignored settings you can put values you expect into comments up there.
3. On a new machine, copy the setting and a commented line

**Comment by Natetronn** [comment]
@celynwalters I'm not sure I'm following, would you mind elaborating or sharing some code please?

**Comment by celynwalters** [comment]
I have these lines like these at the top of my `settings.json`, which are synced everywhere.
When I sync to a new Linux/Windows/Mac machine, I copy each setting _key_ to the main settings body, with a commented line being a suggestion for its _value_. Because each setting is not synced you don't have to keep changing it to fight the synchronisation, just set once and forget.
```json
{
	"settingsSync.ignoredSettings": [
		"editor.fontFamily",
		// CaskaydiaCove Nerd Font Mono
		// CaskaydiaCove NF
		"python.pythonPath",
		// /usr/bin/python3
		// C:\Users\celyn\AppData\Local\Microsoft\WindowsApps
		// /usr/local/bin/python3
		"python.condaPath",
		// /scratch/Code/conda
		// C:\\ProgramData\\miniconda3\\condabin\\conda
		// /Users/celyn/miniconda3/condabin/conda
	],
```

**Comment by Natetronn** [comment]
Thanks for the reply. 

I see the Settings Sync on Windows but, not seeing it on Linux (Manjaro.) It's v1.48.2 and there isn't an update in AUR but, I'll see if I can update manually somehow. 

https://code.visualstudio.com/docs/editor/settings-sync

While researching Setting Sync, I found this article about the Setting Sync extension, which seems similar to the built in preview version, it would seem. I'll leave it here in case it's helpful:

https://zellwk.com/blog/sync-mac-windows-vscode/

**Comment by yzhang-gh** [comment]
I can see that this issue was opened long before we had the official _Settings Sync_ feature. But as it is tagged with `settings-sync` now, I'd like to raise a discussion in relation with #92393 (allowing to sync multiple settings profiles).

IMHO, this feature (settings by platform) is in the subset of #92393 (multiple profiles). There may also be cases that users need multiple profiles even for a same platform (e.g., for work and for personal devices).

**Comment by Natetronn** [comment]
@yzhang-gh it would be great if this was built in! Most of the solutions I've found have been less than ideal.

**Comment by wolf99** [comment]
@yzhang-gh Totes, implementation of the feature request on #92393 would implicitly encompass this use case I believe

**Comment by aidan-fitz** [comment]
> @yzhang-gh it would be great if this was built in! Most of the solutions I've found have been less than ideal.

Same. For my current project, we have a `settings.json` file tracked in the Git repository and need it to have different `python.pythonPath` values for each OS. I'd like to be able to use one settings file across multiple platforms with no extensions.

**Comment by chri7325** [comment]
@aidan-fitz we have the the exact same problem. The same goes for the python linter and formatter path. Our workaround is to use unix as the default and then have the Window's developers set their global configuration for those but it is confusing at first.

**Comment by H-G-Hristov** [comment]
I just run into that problem. We have /src directory that is shared with a docker container which mounts the same directory. It turns out that on the host .vscode is the same directory as seen in the container so I cannot use VSCode for both. Maybe the best option would be to have a subfolder for each platform. So that all files launch.json, c_cpp_properties.json can be independent.

**Comment by brunoamancio** [comment]
Happy birthday, issue 5595! I wish you success and hope to see you again in 10 years, when you are closed.

**Comment by Tyriar** [comment]
New requirement for this from https://github.com/microsoft/vscode/issues/45444, we would still want platform-specific defaults for this ideally that would be filled in. ie. say you autocompleted:

```json
[isWindows]: {
  "terminal.integrated.profiles"
}
```

It should fill in the Windows-specific defaults (git bash, cmd, powershell).

**Comment by iamwwc** [comment]
@Tyriar When this feature will be in next release :), 5 years old. I try use "[isWIndow]":{}, but don't works

**Comment by MostHated** [comment]
Agreed, this is increasingly and ridiculously necessary.

**Comment by Tyriar** [comment]
The deal 👇 

![image](https://user-images.githubusercontent.com/2193314/114792076-8337c180-9d3c-11eb-82cc-74dfc4bf81e4.png)

I don't believe this is on the 2021 roadmap, keep voting though as that increases the visibility of the issue.

**Comment by 1raghavmahajan** [comment]
> keep voting though as that increases the visibility of the issue.

Could you point out where should we vote?

**Comment by 1raghavmahajan** [comment]
> keep voting though as that increases the visibility of the issue.

@Tyriar Could you point out where should we vote?

**Comment by MmAaXx500** [comment]
@1raghavmahajan On the first post with a 👍

**Comment by bonassifabio** [comment]
Please address this request. It is so frustrating that each time I change my computer I need to change the Python Path!

**Comment by alo0oz** [comment]
Still waiting for this................

**Comment by KarFan-FS** [comment]
+1

**Comment by Sv3rk3r** [comment]
This must be a fundamental functionality that really needs to be implemented soon!

**Comment by grrowl** [comment]
Would love to see this feature, ideally similar to how `colorCustomisations` are specified:

```json
{
  "workbench.sideBar.location": "left",
  “[web]”: {
      "workbench.sideBar.location": "right",
  }
}
```

**Comment by JRHeaton** [comment]
I use a small Mac laptop and a huge dual-screen Windows setup at different times, it would be very nice to be able to apply settings per-platform, such as font size, and scrolling speeds.

**Comment by gvcallen** [comment]
@pouyakary with you on that one. any extension with a setting which is directory-based essentially cannot have that setting synced due to this issue. :( one can hope

**Comment by pouyakary** [comment]
@gvcallen sorry I don't understand what are you mentioning. Can you please explain a bit more?

**Comment by MostHated** [comment]
```json

// Can't sync this setting, because many of us use both Linux, Windows, and Mac for Work or Home
// and if we do sync it, it will be wrong on other PC's that are different Operating System
{
    "customExtension.requiredPath": "/path/that/is/needed",
}

```

This is what we need. A way to define OS-specific values to necessary parameters
```json

// 
{
	"if.os.Windows":[
		"customExtension.requiredPath": "C:\\User\MyUer\path",
	],	
	"if.os.Linux":[
		"customExtension.requiredPath": "/home/user/path",
	],
    
}

```

Similar to how in C# you can use compiler definition so that you can use the same variable names and not break code
```cs
#if Windows
	var myPath = "";
#elseif Linux
	var myPath = "";
#else Macos
	var myPath = "";
#endif
```

or like this:

```cs
using System.Runtime.InteropServices;

public static class OperatingSystem
{
    public static bool IsWindows() =>
        RuntimeInformation.IsOSPlatform(OSPlatform.Windows);

    public static bool IsMacOS() =>
        RuntimeInformation.IsOSPlatform(OSPlatform.OSX);

    public static bool IsLinux() =>
        RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
}

// Then this can be used:

if (OperatingSystem.IsLinux())
{
    Console.WriteLine("This code works only on Linux");
}

if (OperatingSystem.IsWindows())
{
    Console.WriteLine("This code works only on Windows");
}

```

**Comment by Natetronn** [comment]
I hate to mention this plugin, because I want to see this feature come to fruition and I don't want its existance (or others similar, which I could never get to work) to be used as an excuse for such a needed FR, but I thought some of you might find it useful, at least while you wait: 

https://marketplace.visualstudio.com/items?itemName=runarsf.platform-settings

**Comment by MostHated** [comment]
Like you, I have tried that plugin, and any others I could find, and the hassle, or often overall ineffectiveness of them are why I am here once again bumping this thread.

**Comment by gvcallen** [comment]
@pouyakary essentially, if I have an extension that has a setting which is a folder path (let's say, a directory to a bunch of project templates), then I cannot successfully sync this setting without issues, because folder paths are fundamentally platform independent. Overall, any setting storing a folder path has to be synced manually

**Comment by pouyakary** [comment]
@gvcallen I think you've confused me for someone else. I don't contribute to the extension settings of the vscode. I though somehow it can be related to my extensions but I don't have any extensions that have user settings.

**Comment by tishion** [comment]
I really need this feature. I always work on C/C++ projects for cross-platforms with CMake and I only keep one source repo on my machine and share them in different platforms.
e.g.  
1. I keep the project source on my windows machine and access the source in WSL
2. when I am working in Windows I need to configure the project with CMake then it will generate the build folder at ${workspaceFolder}/build
3. when I am working in WSL with CMake configure it also generate the build folder at ${workspaceFolder}/build
4. and then i go back to Windows I need to reconfigure it with overwriting the ${workspaceFolder}/build folder

In real project, i need to jump in Windows/Linux/macOS, this is really annoying.

if we can have platform specified settings.json, we can config the build folder as follows.
    "cmake.buildDirectory": "${workspaceFolder}/.build/{$platformName}"

**Comment by AbdulAbCellera** [comment]
> I really need this feature. I always work on C/C++ projects for cross-platforms with CMake and I only keep one source repo on my machine and share them in different platforms. e.g.
> 
> 1. I keep the project source on my windows machine and access the source in WSL
> 2. when I am working in Windows I need to configure the project with CMake then it will generate the build folder at ${workspaceFolder}/build
> 3. when I am working in WSL with CMake configure it also generate the build folder at ${workspaceFolder}/build
> 4. and then i go back to Windows I need to reconfigure it with overwriting the ${workspaceFolder}/build folder
> 
> In real project, i need to jump in Windows/Linux/macOS, this is really annoying.
> 
> if we can have platform specified settings.json, we can config the build folder as follows. "cmake.buildDirectory": "${workspaceFolder}/.build/{$platformName}"

One way I deal with this is by using the compiler in the path like cmake.buildDirectory": "${workspaceFolder}/build/{$buildKit}"

I used environment varibles as well to customize this between different OSs.

More variables for you to use can be found here  https://github.com/microsoft/vscode-cmake-tools/blob/main/docs/cmake-settings.md

**Comment by leodutra** [comment]
The app is multiplatform but closed requests about including a config for multiplatform.
"Interesting."

**Comment by tishion** [comment]
> > I really need this feature. I always work on C/C++ projects for cross-platforms with CMake and I only keep one source repo on my machine and share them in different platforms. e.g.
> > 
> > 1. I keep the project source on my windows machine and access the source in WSL
> > 2. when I am working in Windows I need to configure the project with CMake then it will generate the build folder at ${workspaceFolder}/build
> > 3. when I am working in WSL with CMake configure it also generate the build folder at ${workspaceFolder}/build
> > 4. and then i go back to Windows I need to reconfigure it with overwriting the ${workspaceFolder}/build folder
> > 
> > In real project, i need to jump in Windows/Linux/macOS, this is really annoying.
> > if we can have platform specified settings.json, we can config the build folder as follows. "cmake.buildDirectory": "${workspaceFolder}/.build/{$platformName}"
> 
> One way I deal with this is by using the compiler in the path like `"cmake.buildDirectory": "${workspaceFolder}/build/${buildKit}"`. I, also, sometimes, use environment variables to customize this between different OSs. For more variables to use: https://github.com/microsoft/vscode-cmake-tools/blob/main/docs/cmake-settings.md

Hi @AbdulAbCellera thanks for sharing the useful experience, I will have a try!

**Comment by XuehaiPan** [comment]
I need this feature too to sync my settings across devices (cross platforms). There are many differences between Windows and Unix-like systems (macOS / Linux, etc.).

- The path separators (`/` on Unix-like and `\` on Windows) and delimiters (`:` on Unix-like and `;` on Windows).
- The same program have different configurations on different platforms. For example, I need to set:
    - Linux: `"python.condaPath": "~/Miniconda3/condabin/conda"`
    - Windows: `"python.condaPath": "~\\Miniconda3\\Scripts\\conda.exe"`

  for my Python extension. Same situations for `JAVA_HOME` and `GOPATH`.
- Incompatible font families. I'm using nerd fonts for my customized shell themes. On Linux, I need to set `'DejaVuSansMono Nerd Font Mono'` but `'DejaVuSansMono NF'` on Windows.

**Comment by fredrikaverpil** [comment]
@XuehaiPan can you try setting e.g. an `CONDA_EXEC_PATH` environment variable on your systems and then use `"python.condaPath": "${env:CONDA_EXEC_PATH}`?

**Comment by XuehaiPan** [comment]
@fredrikaverpil it works but I don't think `vscode` (version 1.62.2) supports environment variable substitution in settings yet. See #2809.

I install `conda` in my home directory, and the Python extension will search for the `conda` candidate at startup (function `getCandidatesFromKnownPaths()`).

https://github.com/microsoft/vscode-python/blob/3698950c97982f31bb9dbfc19c4cd8308acda284/src/client/pythonEnvironments/discovery/locators/services/conda.ts#L109-L118

Actually, if you install `conda` in your home directory, you do not need to set `"python.condaPath"` in your settings.

------

This is a special case the extension handles. I still want `vscode` support scope settings by platform feature.

**Comment by Absolucy** [comment]
Here's my use-case for this:

```json
"windows": {
	"rust-analyzer.checkOnSave.extraArgs": [
		"--target-dir",
		"C:\\Users\\Lucy\\AppData\\Local\\Temp\\rust-analyzer-check"
	],
},
"linux": {
	"rust-analyzer.checkOnSave.extraArgs": [
		"--target-dir",
		"/tmp/rust-analyzer-check"
	],
},
"macos": {
	"rust-analyzer.checkOnSave.extraArgs": [
		"--target-dir",
		"/Users/Lucy/Library/Caches/TemporaryItems/rust-analyzer-check"
	],
}
```

**Comment by OgulcanCelik** [comment]
my use case is like this:

To use specific fonts for editor and terminal, you have to use different values for font families.

i.e. to use Patched Hack Font, on WIndows I need to use ` "editor.fontFamily": "Hack NF Regular", but on Linux, I need to use `"editor.fontFamily": "Hack NF Regular"`.  I mean it's 2022 guys.

**Comment by philipbel** [comment]
A little over two months until the issue's 5th birthday. Can't wait 🥳

**Comment by H-G-Hristov** [comment]
launch.json by platform :)

**Comment by ThatGuyRussell** [comment]
@OgulcanCelik I have been able to get around the font family issue with:
`"editor.fontFamily": "Hack NF, Hack Nerd Font"`

Since `Hack NF` isn't on my Mac it correctly falls back to `Hack Nerd Font`

**Comment by JRHeaton** [comment]
This issue feels like it has been up for a ridiculously long amount of time for how direct and obvious of a necessary feature this is for multi platform users. If you make a multi platform app, you must understand that not all software or settings are compatible. As developers I’d figure this is obvious. Enough examples have been provided to demonstrate the need for this and I would like to know from somebody on the collaboration team or main code team what is the hold up or concern in implementing this feature? It’s long overdue.

**Comment by kaem-e** [comment]
they're never going to address this are they

**Comment by tautomer** [comment]
The [Settings Sync extension](https://marketplace.visualstudio.com/items?itemName=Shan.code-settings-sync#review-details) can be a workaround. You can have different gists for different platforms. However, since they are diferent gists, I guess it's more like a backup than syncing across devices (for the same os it might be ok though).

**Comment by arsdever** [comment]
Did no one find a workaround? It's been 6years since this issue was reported. I want to specify different `clang-format` executable paths based on the platform since I use the sync settings feature. And I'm not able to do it?

**Comment by nebbles** [comment]
@arsdever 
As mentioned above in https://github.com/microsoft/vscode/issues/5595#issuecomment-564966427, we had the same problem. Since that post (over 2 years ago), we haven't needed to change our solution of using the [platform-settings](https://marketplace.visualstudio.com/items?itemName=runarsf.platform-settings) extension. 

We committed the clang-format binaries that we needed into the repository. Then we configured the platform-settings extension in the `.vscode/settings.json`:
```
"platformSettings.autoLoad": true,
"platformSettings.platforms": {
    "darwin": {
        "nodes": {
            "C_Cpp.clang_format_path": "${workspaceRoot}/bin/clang-format/darwin_x64/clang-format"
        }
    },
    "linux": {
        "nodes": {
            "C_Cpp.clang_format_path": "${workspaceRoot}/bin/clang-format/linux_x64/clang-format"
        }
    },
    "win32": {
        "nodes": {
            "C_Cpp.clang_format_path": "${workspaceRoot}/bin/clang-format/win32/clang-format.exe"
        }
    }
},
```

**Comment by arsdever** [comment]
Thanks for your quick reply @nebbles. I don't like having installed a lot of extensions (the default set for supporting `c++` and `cmake` is quite enough for me 😄 ). I'm thinking to solve the problem by configuring the `PATH` variable to point to the correct place, or, maybe, having a dedicated `envvar` for that and using it in the settings.

It's sad to know that the thread wasn't revisited even having this many vote-ups. It would have provided a ton of flexibility, especially when you have user sync settings.

**Comment by wolf99** [comment]
Possibly this issue should be closed in favor of #116740 so as to "centralize" support and effort ?

**Comment by kankaristo** [comment]
> Possibly this issue should be closed in favor of #116740 so as to "centralize" support and effort ?

That looks like a possible way to *implement support* for per-platform settings, but other than that, it looks like a different (and broader) feature, with a different set of requirements. I would expect per-platform settings to be automatic, instead of having to manually switch profiles.

I'm also not sure how syncing those profiles would work in practice. Would the current profile also be stored in the settings JSON? If so, it would be synced like any other setting, and it would essentially break per-platform settings: computer A should permanently use profile A (e.g. Linux settings), and computer B should permanently use profile B (e.g. macOS settings). Under other circumstances, syncing the used profile across computers might be preferable, but not here.

And as a last minor thing, an issue number of 5595 (vs. the much larger 116740) immediately tells that this is a long-awaited feature. So, until GitHub (finally) has some proper "issue popularity" feature, it would hurt the visibility of this issue, if it were closed in favor of a much more recent issue.

**Comment by nebbles** [comment]
> until GitHub (finally) has some proper "issue popularity" feature

If you count 👍  reactions as upvotes (which many deem the unofficial way of voting/showing support) then sorting by that reaction shows [this issue is 14th in popularity for this repo](https://github.com/microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)

Plus from my view, user profiles are trying to solve a different problem to platform-specific settings. I would imagine there are plenty of people quite content with just one profile, but just may have to use multiple platforms. Or others (in the case I outlined above) where the vscode settings are shared with the repo and we have contributors using different platforms that should have the repo automatically configured to work for them out the box.

**Comment by medo64** [comment]
Due to different font rendering between platforms (in my example Windows 10 and Ubuntu 22.04), for some fonts I need a different font weight to get the same look and feel. Having per-platform fontWeight (and, god-forbid, other font related settings) would solve this problem neatly.

But I doubt this corner case will move the scales when other examples in the previous 5 years didn't...

**Comment by cwensley** [comment]
This has been a bane of our existence enough times that it made me have to comment on another use case with _no workaround_.  We use the C# extension and have different .sln's to use based on platform.    On the other platform(s) we need to manually select the project every time, which is simply a waste of time..

```
{
    "omnisharp.defaultLaunchSolution": "path/to/windows.sln", // on Mac or Linux you need to select the correct sln every time.
}
```

Please, please implement this feature.  We can't have all plugins implement a way to have different settings per platform, it is wasted effort when it can be implemented globally.

**Comment by matrix-ranger** [comment]
I need this feature, too. Please implement it soon. It should be implemented globally as with `per platform keybindings`.

**Comment by thernstig** [comment]
@sandy081 @kieferrm in 2016 User Voice was retired. Github reactions replaced this functionality to gauge user interest of new features.

https://code.visualstudio.com/blogs/2016/08/19/goodbyeuservoice

This is voted top 12 of all users requests currently. Is there any chance prioritization could honor the use of reactions, for issues that I _guess_ are fairly straight forward to implement? Why this specific issue is important is that it is a real problem for many users, not just a "nice to have".

**Disclaimer**
I understand asking for prioritization considerations can be frowned upon and if so, I preemptively apologize. You are doing an amazing job nonetheless I and do understand if you prioritize otherwise. After all, most of us are developers that understand prioritization has nuances not always visible to external users.

**Comment by JRHeaton** [comment]
This is super silly to be continuously asking at this point. It's likely trivial for the core devs and I speculate the only reason it's de-prioritized or ignored outright is that they're all using a single platform (Windows). I think we should dupe this issue multiple times to gain traction on it. Just like with Apple's radar/bug reporting system where you can refer to an existing issue/ticket and dupe it, I think we should do the same on GitHub here so that this gets implemented. Otherwise I'm waiting, and have been for over a year, to hear the rationale from the dev team as to why they would purposely **not** implement it. At least be transparent about receptoin and potential timelines and priorities.... anyone else with me?

**Comment by ooobii** [comment]
Not sure why this isn't being worked on. How can we escalate this issue to the developers? Especially when switching between Windows & Unix environments where path specifications are different, this is a seriously troubling issue...

**Comment by piotrku** [comment]
I came here today and I'm kind of surprised that so many people has this problem for so long and nothing happens, and even more that solving it does issue not require any significant amount of work :(

**Comment by AbdulAbCellera** [comment]
Sorry to spam everyone but could this solve our issue? https://code.visualstudio.com/updates/v1_69#_settings-profiles

**Comment by nickmccurdy** [comment]
Is there a way to automatically set a profile based on the platform?

**Comment by michaelmarziani** [comment]
I have a similar use case with a mixed dev environment. This feature would save a good bit of confusion and frustration.

**Comment by Commybwans** [comment]
This is a great feature if vscode can support it, now I have to ues different github users.

**Comment by sanyer** [comment]
@AbdulAbCellera this looks exactly like a perfect solution and even more!

**Comment by deividrvale** [comment]
I'm searching around on how to solve the exact same problem.
But I need to set some paths on the .json config file... and you probably guessed it right if you guess was that they are different between Linux/Mac/Windows.

It seems the feature isn't implemented yet even after so many years.

Very sad.

**Comment by Cielquan** [comment]
This feature seems to be implemented as beta in the insiders version since last month. You may want to have a look at that.

See comment above: https://github.com/microsoft/vscode/issues/5595#issuecomment-1178497520

> Sorry to spam everyone but could this solve our issue? https://code.visualstudio.com/updates/v1_69#_settings-profiles

**Comment by mikew** [comment]
I don't think Profiles really solve this. It copies your whole settings and then keeps them separate. So for me, who has 200 odd lines of settings across 3 platforms, it's now 600 lines of settings overall. And if you want a change to apply to all profiles, you have to manually switch to it, change it, then change back.

I just want to say "Windows should use native title bars", not manage and switch between _N_ separate profiles. Seems like this should be the interface for each VS Code settting:

```ts
type VSCodeSetting<T> = T | {
  windows: T
  osx: T
  linux: T
}
```

**Comment by deividrvale** [comment]
> I don't think Profiles really solve this. It copies your whole settings and then keeps them separate. So for me, who has 200 odd lines of settings across 3 platforms, it's now 600 lines of settings overall. And if you want a change to apply to all profiles, you have to manually switch to it, change it, then change back.
> 
> I just want to say "Windows should use native title bars", not manage and switch between _N_ separate profiles. Seems like this should be the interface for each VS Code settting:
> 
> ```ts
> type VSCodeSetting<T> = T | {
>   windows: T
>   osx: T
>   linux: T
> }
> ```

This describes exactly what I wanted.

**Comment by autoferrit** [comment]
don't get me wrong, the new profiles feature is pretty nice. But I agree that it copying the config is really cumbersom. why not have a base config and the other profile has a parent it inherits from and only implements the changed config items? That alone would probably make profiles work. 

Beyond that, I have to tell vscode settings sync to ignore completely the important config items. which isn't a small amount. At least theres specific platform config for some items, but those are mostly because plugin authors have made those per platform config items, not native to vscode.

Granted I think 
> bloated product and yet can't do basic things like per os config, for a crossplatform tool, shame on microsoft
is a bit much as it seems like most editors have some kind of issue with cross platform configs, even the IntelliJ stuff. Mostly that's not as much of an issue with terminal-based editors like vim and emacs. But those are a different ballgame anyways. 

One idea, would be that if any config was just prefixed with the platform, or hostname, then it would only apply on that one platform/hostname. but using the blocks like `[python]` -> `[darwin]` would be far better

**Comment by xenoterracide** [comment]
honestly, I would like a per computer setting, nothing guarantees a path, for example, will be the same if I switch computers even if they are on the same platform.

**Comment by julianxhokaxhiu** [comment]
Came here for the very same request. Kindly asking if it's possible to be considered by the VSCode team. Or if  possible, let me know of how can I support different settings on different platforms, for eg. [CMake Tools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools).

**Comment by scmx** [comment]
You could generate the file to be correct for the platform. Here's one specific way for a use-case I had.
> VSCode settings.json `typescript.tsdk` double backslash windows workaround Next.js
> https://gist.github.com/scmx/de7564c9efebaec108cfb74e83ec9d1b

**Comment by julianxhokaxhiu** [comment]
@starball5 thanks, in the end I eventually migrated to CMake Presets on the project i was requiring this and now I'm unblocked. Also this seems the way to go, so it's not only VSCode related but it supports the entire CMake ecosystem :)

**Comment by emmenlau** [comment]
@assiless I'm not sure what you are saying. Does this work now?

**Comment by znorman-harris** [comment]
As an extension developer, minimally we should have the ability to set platform, just like scope, for preferences in package.json files. Maybe something like this:

![image](https://user-images.githubusercontent.com/31664668/230643623-dd7d60f3-c768-4987-b9b4-d8674f953a7e.png)

**Comment by Poikilos** [comment]
Here's a good use case: Using ShellCheck shows a linting error when the file contains "\r" characters (such as in Windows which uses "\r\n", or macOS <= 9 which uses "\r"), but git is usually (such as in GitHub Desktop and Git Cola) set to check out newlines using the platform's newlines but commit using "\n" (Therefore there isn't actually lint). Therefore to prevent Windows users from seeing every line as having an error in ShellCheck, the following setting should only be for win32:
```
"shellcheck.exclude": ["1017"],
```
- If set for every OS, it could potentially hide actual lint on macOS and Linux-based systems. Therefore the linting error should only be excluded on win32.

**Comment by puddlejumper9** [comment]
ChatGPT-4 just hallucinated this feature by telling me to use platform specific settings in settings.json like in launch.json:
"osx": { ... }

This is a brilliant idea and since the OpenAI terms of use give me full ownership over the answers I am taking credit for it... as reinventing this independently since I found it was already asked for.

We have it in launch.json, why not ALL .vscode/*.json files???

**Comment by gueggel** [comment]
Any news? This would be a great and really helpful feature.

**Comment by pouyarezvani** [comment]
Just use Intelij guys 😁

**Comment by philipbel** [comment]
Well, it's been 7 years, 5 months and 25 days since this issue was posted.

**Comment by ievgennaida** [comment]
Would be nice to have python venv based on envrionment:
`"python.defaultInterpreterPath": "~/venv/bin/python"`

**Comment by thernstig** [comment]
> Would be nice to have python venv based on envrionment: `"python.defaultInterpreterPath": "~/venv/bin/python"`

This is an aside, but I recommend you use:

```
"python.defaultInterpreterPath": "${workspaceFolder}/.venv/bin/python",
```

**Comment by TryerGit** [comment]
This will be a huge time saver and boost to productivity. [Settings sync](https://code.visualstudio.com/docs/editor/settings-sync) does not solve the issue. See stackoverflow discussion on that [here](https://stackoverflow.com/questions/74348747/settings-sync-settings-json-is-different-between-wsl-and-windows-on-same-mach)

Currently, I have to manually manage the following:

```
Windows Machine 1:

    "hsnips.hsnipsPath": "C:\\Users\\TryerGitOne\\AppData\\Roaming\\Code\\User\\hsnips",

Windows Machine 1, WSL: 

    "hsnips.hsnipsPath": "/mnt/c/users/TryerGitOne/AppData/Roaming/Code/User/hsnips",

```
```
Windows Machine 1:

    "vim.neovimPath": "C:\\Users\\TryerGitOne\\Downloads\\nvim-win64\\Neovim\\bin\\nvim.exe",
    "vim.vimrc.path": "C:\\Users\\TryerGitOne\\.vimrc",
    "hsnips.hsnipsPath": "C:\\Users\\TryerGitOne\\AppData\\Roaming\\Code\\User\\hsnips",

Linux: 

    "vim.neovimPath": "/usr/bin/nvim",
    "vim.vimrc.path": "/home/TryerGitOne/GoogleDrive/research_programming/Setup Folder/VSIDE Stuff/.vimrc",
    "hsnips.hsnipsPath": "~/.config/Code/User/hsnips",

```

```
Windows Machine 1:

    "vim.neovimPath": "C:\\Users\\TryerGitOne\\Downloads\\nvim-win64\\Neovim\\bin\\nvim.exe",
    "vim.vimrc.path": "C:\\Users\\TryerGitOne\\.vimrc",
    "hsnips.hsnipsPath": "C:\\Users\\TryerGitOne\\AppData\\Roaming\\Code\\User\\hsnips",

Windows Machine 2:

    "vim.neovimPath": "C:\\Users\\TryerGitTwo\\Downloads\\nvim-win64\\bin\\nvim.exe",
    "vim.vimrc.path": "C:\\Users\\TryerGitTwo\\.vimrc",
    "hsnips.hsnipsPath": "C:\\Users\\TryerGitTwo\\AppData\\Roaming\\Code\\User\\hsnips",
```

```
Windows Machine 1:

    "vim.neovimPath": "C:\\Users\\TryerGitOne\\Downloads\\nvim-win64\\Neovim\\bin\\nvim.exe",
    "vim.vimrc.path": "C:\\Users\\TryerGitOne\\.vimrc",
    "hsnips.hsnipsPath": "C:\\Users\\TryerGitOne\\AppData\\Roaming\\Code\\User\\hsnips",

Windows Machine 2, WSL: 

    "vim.neovimPath": "C:\\Users\\TryerGitTwo\\Downloads\\nvim-win64\\bin\\nvim.exe",
    "vim.vimrc.path": "C:\\Users\\TryerGitTwo\\.vimrc",
    "hsnips.hsnipsPath": "/mnt/c/users/TryerGitTwo/AppData/Roaming/Code/User/hsnips",
```

You can easily see the combinatorial madness. Many of these machines (even Windows machines) are official machines where the user does not have the option of providing the same username as all other machines so that the directory paths may be the same. 

Vim already provides for this in `.vimrc`. Here is an example from my `.vimrc` where based on running a shell command (query present working directory), one can figure out which machine mount one is on:

```
let g:machine_run_on_wsloffice = system("pwd | grep -c '/mnt/e'")
let g:machine_run_on_wslhome = system("pwd | grep -c '/mnt/c'")
if g:machine_run_on_wsloffice == 1
  set viminfo=%,<800,'100,/50,:100,f1,n./.vim/.viminfooffice
else
  if g:machine_run_on_wslgm == 1
    set viminfo=%,<800,'100,/50,:100,f1,n./.vim/.viminfohome
  else
    set viminfo=%,<800,'100,/50,:100,f1,n./.vim/.viminfoubuntu
  endif
endif
```

As others have pointed out, `CMake` has support for something similar also.

**Comment by inu1255** [comment]
If you have a Mac and a Windows, and some configurations are different, you can ignore these configurations by setting `settingsSync.ignoredSettings`

**Comment by tn-072** [comment]
What about this:

```jsonc
{
    // ......
    "[sys:win]": {
            "python.defaultInterpreterPath": "<my python path on windows>",
    },
    "[sys:linux]": {
            "python.defaultInterpreterPath": "<my python path on linux>",
    },
    "[sys:osx]": {
            "python.defaultInterpreterPath": "<my python path on Mac>"
    },
}
```

**Comment by sdholden** [comment]
Some sort of solution is long overdue...any progress?

**Comment by buzcarter** [comment]
I'll second @tn-072 's suggestion. 

```json
{
    "editor.fontSize": 14,
    "python.pythonPath": "/usr/local/bin/python3",
    "[windows]": {
        "editor.fontSize": 16,
        "python.pythonPath": "C:\\Python39\\python.exe"
    },
    "[mac]": {
        "editor.fontSize": 18,
        "python.pythonPath": "/usr/local/bin/python3"
    },
    // ... the shared bits
}
```

**Comment by sanmai-NL** [comment]
Instead of scoping by 'platform' I suggest to scope by user-defined tag(s), to support use cases where the same OS is used for different purposes (e.g. work Linux, personal Linux) beyond what profiles can offer.

**Comment by sanmai-NL** [comment]
Then all combinations of language, macOS/Linux/Windows, work/personal world need to be separate profiles.

**Comment by starball5** [comment]
I ctrl+f-ed for "profile" and several times when the profiles approach came up in past discussion, it was dismissed because at the time, it required creating a standalone copy of settings, and there was no way to inherit from the default. For those people, see #156144, which as released in July 2023.

**Comment by jd-solanki** [comment]
Why it's still in backlog?

**Comment by pfeerick** [comment]
I'd love to see this feature for the some of the above reasons such as OS and path specific differences, but also because I set the new option
```json
"terminal.integrated.middleClickBehavior": "paste"
```
which works great on Windows, but on Linux (which picks up the setting because I have Settings Sync enabled), it makes it paste the clipboard content + selection rather than just the selection, so I've had to set it back to default now. :rofl:

**Comment by starball5** [comment]
@pfeerick oh interesting. Are you aware of `"editor.selectionClipboard": false,`? See also https://stackoverflow.com/q/36723333/11107541

**Comment by Poikilos** [comment]
> Then all combinations of language, macOS/Linux/Windows, work/personal would need to be separate profiles.

If you have a work and personal account on your linux/macOS/Windows machine, that is 6 accounts. If you need it that fine-grained, you could already just put the setting in your user profile. It isn't directly related to this. There is no way to detect if you are work or personal...that is a user-level abstract distinction.

**Comment by pfeerick** [comment]
> @pfeerick oh interesting. Are you aware of `"editor.selectionClipboard": false,`? See also 

No I wasn't... thanks for that!. It fixes the "double" paste (one from selection, one from clipboard), but does mean you must use the clipboard if using `middleClickBehavior": "paste"`, so still not ideal. Spotted `terminal.integrated.copyOnSelection` as well though which I didn't know about... that would work if just working with terminal, and not wanting to mix editor and terminal.

**Comment by alexchexes** [comment]
In the meantime (the discussion has lasted 8 years and doesn't seem to be progressing), how do you all set a comfortable font when switching machines? On my MacBook, I'm good with a 12px font, but on my Windows machine, I need at least 14px. What do you do in such cases?

**Comment by sanmai-NL** [comment]
You can use settings profiles nowadays. Maybe you can template them?

**Comment by alexchexes** [comment]
@sanmai-NL
Could you please provide a bit more detail about your suggestion? I checked out the VSCode documentation, but it seems that templates refer to a predefined set of configurations used for different programming languages or common environments. Not sure how I could use that to keep all my settings in sync while having different font sizes on Mac and Windows...

**Comment by sanmai-NL** [comment]
I mean templating VS Code settings with an external tool, and putting the per-OS outputs in a dotfiles repository.

**Comment by hermes85pl** [comment]
> What do you do in such cases (when Settings Sync is On)?

@alexchexes, one solution would be to treat such settings as device-specific and put them under `settingsSync.ignoredSettings`.

**Comment by iwangbowen** [comment]
Any updates

**Comment by rnetonet** [comment]
Please, this will be so useful!

**Comment by JamesSwivel** [comment]
I guess platform-specific settings are useful. For example, I want to have one settings.json that can address Windows and Linux Python virtualenv requirements.  For example below
```json
{
  "windows": {
     "python.defaultInterpreterPath": ".\\winEnv\\Scripts\\python.exe",
  },
  "linux": {  
      "python.defaultInterpreterPath": "${env:HOME}/.pyenv/versions/3.11.8/envs/virtualenvX/bin/python3",

}
```
Any suggestion to address this common Python requirement?

**Comment by Hirak232** [comment]
[Advertisement-JW-253-posts-260123.pdf](https://github.com/user-attachments/files/16844199/Advertisement-JW-253-posts-260123.pdf)

**Comment by eeplus-oka** [comment]
When standardizing environment settings across project teams, I think the easiest way is to distribute the settings.json along with the source code. How about splitting the files in a similar way to Visual Studio's appsettings.json (e.g., appsettings.development.json and appsettings.production.json)?

- settings.json
- settings.[platform].json
- settings.windows.json
- settings.macos.json

**Comment by iwangbowen** [comment]
We need this feature now

**Comment by jd-solanki** [comment]
Even tough this issue has lots of upvotes why it's in backlog? @sandy081

**Comment by workingj** [comment]
Yeah, would be really great to have!

**Comment by josephgarnier** [comment]
An alternative solution would be for _Settings_ to support [variable substitution](https://code.visualstudio.com/docs/editor/variables-reference), even if it means adding other variables.

Implementing another type of settings (by OS) could become cumbersome to manage, given that we already have settings by scope (Global, Workspace, Folder), plus settings by profile.

**Comment by dendobi** [comment]
I hope Santa Claus brings me this feature this year.

**Comment by Tucker6742** [comment]
The best answer to this for now is to just make a new profile for each platform and selected the right one for each machine. I have a Windows profile for my home machine and a Linux profile for my work machine, works just fine. You can copy all setting for one profile to another so you only need to change what are different from 2 profile and don't have to setup everything again

**Comment by rrmistry** [comment]
> The best answer to this for now is to just make a new profile for each platform and selected the right one for each machine. I have a Windows profile for my home machine and a Linux profile for my work machine, works just fine. You can copy all setting for one profile to another so you only need to change what are different from 2 profile and don't have to setup everything again

Would still be good to scope settings per profile (instead of by platform as OP requested)

E.g. Having a base/default profile that contains most common settings, and other profiles override with specific settings for their profile scope

**Comment by DaanRosendal** [comment]
Please add this feature 🙏

**Comment by iammohsinar** [comment]
Team make some progress and add this features as I would like to sync settings with my Mac and Windows system.

**Comment by gueggel** [comment]
I think we're just unlucky here. Instead of bringing new, important features or fixing bugs, they're just driving Copilot through the village. Copilot here, Copilot there. It's getting annoying.

**Comment by joseph423** [comment]
Are we still...waiting....for this?

Keep on waiting???!!!

**Comment by Penguinang** [comment]
We need this feature

**Comment by repentsinner** [comment]
We're going to have full agentic coding balogna that doesn't care about
cross-platform compatibility before this feature drops.

**Comment by Joedmin** [comment]
I need this feature because vscode in the web (GitLab Web IDE) uses differents paths in for JSON Schema configs in the [YAML Extension](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml).

Desktop `settings.json`:
```json
  "yaml.schemas": {
    "my.schema.json": "*.yml",
  },
```

vs the `settings.json` web version needs:
```json
  "yaml.schemas": {
    "schemaservice://combinedschema/my.schema.json": "*.yml",
  },
```

And I cannot have both defined at one because there's `Multiple JSON Schemas used to validate this file, click to select one` error...

**Comment by printfdebugging** [comment]
since they are not listening, how about creating an extension which looks for "unix.json" "macosx.json" "windows.json"  and based on the platform you are on, symlinks the appropriate config file to settings.json? or maybe symlinking is not required, one can just source that config?

**Comment by printfdebugging** [comment]
there's an extension which makes it possible, ` runarsf.platform-settings` is the extension id. please see https://github.com/printfdebugging/learn-opengl/blob/main/.vscode/settings.json where i use it, works quite well across windows, linux and mac.

**Comment by kankaristo** [comment]
@printfdebugging, that extension might fix the issue for some people, but unfortunately that extension just modifies `settings.json` on the fly.

This still marks the file as changed in Git (I have `settings.json` in my dotfiles repo), making "actual" changes to settings difficult to keep track of.

If this were an actual feature in VSCode, `settings.json` would **only be read**, not changed on the fly based on the platform.

**Comment by blueben** [comment]
If this is resolved before the issue turns 10 years old, I will drive over to Redmond and take the developer who implements this out for lunch.

**Comment by alexchexes** [comment]
Love round digits, btw:

<img width="471" height="127" alt="Image" src="https://github.com/user-attachments/assets/8f5182fe-e0b0-4b38-9733-5f196f57f993" />

**Comment by emmenlau** [comment]
> I thought this was implemented because you can configure env variables based on OS.

Not for all settings.

**Comment by SEsnorre** [comment]
It is now open for 10 years. Maybe in another 10 we get an implementation. Jokes aside, is there any progress on this topic? Me and a lot of people I know would love this feature.

**Comment by yCodeTech** [comment]
Tbh, with vscode's new release cycle they SHOULD be able to get through a lot more issues than they currently do. But perhaps they're priorities are not aligned with the users at the moment, which is a massive shame

**Comment by sysescool** [comment]
Just as `tasks.json` allows you to specify different commands for different platforms—"windows", "linux", and "osx"—`settings.json` should support this as well. This way, you would only need to keep a single `settings.json` file in sync.

**Comment by lrtfm** [comment]
An implementation is now available for review in #335281.

It adds `windows`, `osx`, and `linux` blocks to `settings.json`. Common settings are applied first, followed by the matching platform block:

```jsonc
{
    "editor.fontSize": 16,

    "osx": {
        "editor.fontSize": 14,
        "[typescript]": {
            "editor.defaultFormatter": "esbenp.prettier-vscode"
        }
    },

    "windows": {
        "editor.fontSize": 15
    },

    "linux": {
        "editor.fontSize": 13
    }
}
```

The implementation also supports nested language-specific settings, schema completion and validation, existing configuration-scope precedence, and Settings Sync.

Unofficial Code - OSS preview builds are available for testing:

- macOS ARM64
- Windows x64 portable
- Windows x64 installer

https://github.com/lrtfm/vscode/releases/tag/os-settings-preview-fe76090f0

These builds correspond to commit [`fe76090f0`](https://github.com/lrtfm/vscode/commit/fe76090f08eefd38bf90bf4328c09d301e44b64a). The macOS build is ad-hoc signed but not notarized, and the Windows packages are unsigned, so the operating system may display a security warning.

Testing and feedback on #335281 would be very helpful.
